src/test/java/org/elasticsearch/index/store/ExceptionRetryTests.java

```diff
@@ -0,0 +1,144 @@
+import org.elasticsearch.transport.*;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.ExecutionException;
+
+import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
+import static org.elasticsearch.index.query.QueryBuilders.termQuery;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.*;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+
+@ElasticsearchIntegrationTest.ClusterScope(scope = ElasticsearchIntegrationTest.Scope.SUITE)
+public class ExceptionRetryTests extends ElasticsearchIntegrationTest {
+
+    volatile boolean exceptionThrown = false;
+
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal) {
+        return ImmutableSettings.builder()
+                .put(super.nodeSettings(nodeOrdinal)).put("gateway.type", "local")
+                .put(TransportModule.TRANSPORT_SERVICE_TYPE_KEY, MockTransportService.class.getName())
+                .build();
+    }
+
+    /**
+     * Tests retry mechanism when indexing. If an exception occurs when indexing then the indexing request is tried again before finally failing.
+     * If auto generated ids are used this must not lead to duplicate ids
+     * see https://github.com/elasticsearch/elasticsearch/issues/8788
+     */
+    @Test
+    public void testRetryDueToExceptionOnNetworkLayer() throws ExecutionException, InterruptedException, IOException {
+        int numDocs = scaledRandomIntBetween(100, 1000);
+        NodesStatsResponse nodeStats = client().admin().cluster().prepareNodesStats().get();
+        List<NodeStats> dataNodeStats = new ArrayList<>();
+        for (NodeStats stat : nodeStats.getNodes()) {
+            dataNodeStats.add(stat);
+        }
+
+        Collections.shuffle(dataNodeStats, getRandom());
+        NodeStats unluckyNode = dataNodeStats.get(1);
+
+        assertAcked(client().admin().indices().prepareCreate("index"));
+        ensureGreen("index");
+
+        //create a transport service that throws a ConnectTransportException for one bulk request and therefore triggers a retry.
+        for (NodeStats dataNode : dataNodeStats) {
+            MockTransportService mockTransportService = ((MockTransportService) internalCluster().getInstance(TransportService.class, dataNode.getNode().name()));
+            mockTransportService.addDelegate(internalCluster().getInstance(Discovery.class, unluckyNode.getNode().name()).localNode(), new MockTransportService.DelegateTransport(mockTransportService.original()) {
+
+                @Override
+                public void sendRequest(DiscoveryNode node, long requestId, String action, TransportRequest request, TransportRequestOptions options) throws IOException, TransportException {
+                    super.sendRequest(node, requestId, action, request, options);
+                    if (action.equals("indices:data/write/bulk[s]") && !exceptionThrown) {
```

Comment:
can this be a constant somehow? I wonder if we have it as a constant already somewhere?